### PR TITLE
Module2

### PR DESCRIPTION
**Pull request by elvyyy**

### CODE REVIEW

File: module_2/controller/src/main/java/com/epam/esm/controller_module/exception/ExceptionResolver.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import com.epam.esm.repository_module.exception.UnknownSearchCriteriaException;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+import javax.validation.ConstraintViolationException;
+
+@RestControllerAdvice
+public class ExceptionResolver {
```
Comment (MariyaBystrova) [inline module_2/controller/src/main/java/com/epam/esm/controller_module/exception/ExceptionResolver.java:17]: I think you need to overthink which exceptions to handle. There are a lot of use cases such as HttpRequestMethodNotSupportedException that would possibly break your application.
+ My recommendation is to add generic handler the the end (Throwable) that will prevent form default error screen with stack trace.

File: module_2/controller/src/main/java/com/epam/esm/controller_module/exception/ExceptionResolver.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler(EntityNotCreatedException.class)
+    public ExceptionInfo resolveEntityNotCreatedException(EntityNotCreatedException exception) {
```
Comment (MariyaBystrova) [inline module_2/controller/src/main/java/com/epam/esm/controller_module/exception/ExceptionResolver.java:20]: Also it is better to log every exception

File: module_2/repository/src/main/java/com/epam/esm/repository_module/repository/BaseDao.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package com.epam.esm.repository_module.repository;
+
+import com.epam.esm.repository_module.domain.BaseEntity;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface BaseDao<T extends BaseEntity> {
```
Comment (MariyaBystrova) [inline module_2/repository/src/main/java/com/epam/esm/repository_module/repository/BaseDao.java:8]: As you have parametrization on BaseEntity, I recommend to parametrize this class with Id type as well and not to hardcode long

File: module_2/repository/src/main/java/com/epam/esm/repository_module/repository/impl/GiftCertificateDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+@Repository
+public class GiftCertificateDaoImpl implements GiftCertificateDao {
+    private static final ResultMapHandler<Optional<GiftCertificate>> handler =
+            ResultMapHandlerFactory.GIFT_CERTIFICATE_RESULT_MAP_HANDLER;
+    private static final ResultMapHandler<List<GiftCertificate>> listHandler =
+            ResultMapHandlerFactory.GIFT_CERTIFICATE_LIST_RESULT_MAP_HANDLER;
+
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
```
Comment (MariyaBystrova) [inline module_2/repository/src/main/java/com/epam/esm/repository_module/repository/impl/GiftCertificateDaoImpl.java:28]: It is better to use autowired on constructor or setter

File: module_2/service/src/main/java/com/epam/esm/service_module/service/TagService.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package com.epam.esm.service_module.service;
+
+import com.epam.esm.repository_module.domain.Tag;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface TagService {
+    Optional<Tag> readById(long id);
```
Comment (MariyaBystrova) [inline module_2/service/src/main/java/com/epam/esm/service_module/service/TagService.java:9]: Same recommendation as for Dao - parametrization for id type

**Review by evgeniyyaskevich — COMMENTED** [review]
Some general words about the review.
First of all, I would like to say thank you, I believe you did a great job.

I would like to notice your "procedure"-based approach to implementation. It seems you are a kind of expert in DB procedures and functions, it's great. DAO classes look really unified. And nevertheless, I am not a fan of such an approach.
Unfortunately, completely not all developers are deeply aware of PL/SQL. It can take significant time to read such procedures and understand the logic. And as a result, it can be difficult to maintain((investigate the issues, fix bugs, make changes) such an application because it can cause unexpected and non-obvious issues.
Such an approach also moves almost all business logic to the database and it doesn't seem a good idea.
I would not like to say that we need to avoid PL/SQL, no. I prefer to write PL/SQL code only if I can't do it via SQL. The general message here is not to complicate the simple operations. It is important to remember about KISS principle.
I hope you practiced also JDBC Template, not only DB procedures (:

Best regards, Yauheni

File: module_2/controller/src/main/java/com/epam/esm/controller_module/controller/GiftCertificateController.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+     * <p>
+     * If there is no gift certificate with the specified id response gets status 404 - Not Found.
+     * The default response status is 200 - OK.
+     *
+     * @param {@code certificateId} Identifier of the requested gift certificate. Inferred from the request URI.
+     * @return {@link GiftCertificate}.
+     */
+    @GetMapping("/{certificateId}")
+    @ResponseStatus(HttpStatus.OK)
+    public GiftCertificate getCertificate(@PathVariable @Positive long certificateId) {
```
Comment (evgeniyyaskevich) [inline module_2/controller/src/main/java/com/epam/esm/controller_module/controller/GiftCertificateController.java:50]: It is also can be reasonable to encapsulate the entities in the repository module. The entity is a representation of the database table row, only DAO should know about it.
I can suggest using DTO(Data Transfer Objects) in the future to reach better isolation of modules.

**Pull request merged**